### ISSUE TITLE
 In YOLOv8, the `Results` object doesn't have a specific attribute for class probabilities (`probs`). 

### ISSUE THREAD

**Issue by cyyuan0819** [ultralytics/ultralytics#5823]
@ABeltramo i apologize for any confusion caused. In YOLOv8, the `Results` object doesn't have a specific attribute for class probabilities (`probs`). Instead, you can access the class confidence scores provided by the model. These scores represent the probability of each detection belonging to a specific class. Please note that these confidence scores are in a raw, unnormalized form (logits), and they are not directly converted to class probabilities. You can use these confidence scores to perform further analysis or apply appropriate post-processing techniques to obtain class probabilities if needed.

I hope this clarifies the situation. If you have any further questions, please let me know.

_Originally posted by @glenn-jocher in https://github.com/ultralytics/ultralytics/issues/2600#issuecomment-1612307619_

**Comment by glenn-jocher** [comment]
@cyyuan0819 in YOLOv8, detections are processed and filtered before they are returned by the model. Each detection includes the objectness confidence, class confidence, and the bounding box coordinates. The class confidence scores are generally high for the predicted class and low for the other classes.

If you want to access the raw class probabilities, you would need to delve deeper into the model's layers. This might involve modifying the model's forward method to return the intermediate outputs before the final non-maximum suppression (NMS) stage. However, do note that these modifications can be complex and may require a good understanding of the code and the underlying YOLO architecture.

As for the calculation of the final confidence score (0.8 in your example), it's actually a combination of objectness confidence and class confidence. YOLOv8 uses an objectness score to represent how likely a bounding box contains an object (any object), and class scores to represent how likely the object belongs to each class. The final confidence score for each class is calculated by multiplying the objectness score with the class score.

Remember to always handle these values carefully. They are at the core of the model's detection mechanism, and any changes can significantly affect the model's performance and results.

**Comment by MoAbbasid** [comment]
@glenn-jocher, Hello, I read several blogs online that says that YOLOv8 ditched the objectness score, judging from your comment YOLOv8 uses an objectness score?
also issues like these arise when there is no official paper, so any news on YOLOv8 paper in the near future?

**Comment by glenn-jocher** [comment]
@MoAbbasid hello! Thanks for reaching out. 😊

Yes, the image you shared correctly indicates that YOLOv8 has evolved from the traditional objectness score approach. Instead of relying on a separate objectness score for each bounding box, YOLOv8 integrates object detection confidence more directly into the model's architecture and output. This change aims to streamline the detection process and improve performance.

If you're diving into the code or working on custom modifications, you'll notice this architectural evolution reflected in how detections are handled and confidence scores are computed.

Happy coding, and feel free to ask if you have more questions!

**Comment by MoAbbasid** [comment]
@glenn-jocher Thank you for your response
can you elaborate more on how object detection confidence is integrated directly into the model's architecture?
for a 2 class problem, the output is now only 6 values instead of 7?

**Comment by glenn-jocher** [comment]
@MoAbbasid Absolutely, happy to clarify! 😊 In YOLOv8, the model's architecture has been refined to directly integrate detection confidence into its predictions. This means that for each bounding box, the model outputs the coordinates (4 values) and the class confidences (in your case, 2 values for a 2-class problem). So, yes, you're right: the output would consist of 6 values per detection instead of 7, as the separate objectness score is no longer explicitly calculated.

This streamlined approach helps in reducing complexity and potentially increases the efficiency of the model. Here's a simplified view of how the output looks for a 2-class problem:

```python
# Example output for a single detection
# [x_center, y_center, width, height, class1_confidence, class2_confidence]
[0.5, 0.5, 0.2, 0.3, 0.8, 0.2]
```

Hope this helps! Let me know if you have more questions.

**Comment by MoAbbasid** [comment]
@glenn-jocher so to confirm, the class1_confidence, and class2_confidence, in your reply are referring to the final confidence ? meaning these values are objectness score x class probabilities?
am I getting that right?

Thanks again for the response

**Comment by MoAbbasid** [comment]
@glenn-jocher Im waiting for an answer so I can proceed, the class1_confidence, and class2_confidences shown in your reply are final confidences that account for objectness (they are objectness score x class probabilities) ?

**Comment by glenn-jocher** [comment]
@MoAbbasid Yes, you've got it right! 😊 In YOLOv8, the values for `class1_confidence` and `class2_confidence` represent the model's confidence that the detected object belongs to class 1 or class 2, respectively. These confidence scores inherently account for the presence of an object (what was traditionally the objectness score) and its class probability. So, there's no separate multiplication step needed; the model directly outputs these integrated confidence scores.

Happy to help if you have more questions!
